### ISSUE TITLE
Debug version of Microsoft.Research.CNTK.CpuEval-mkl does not work

### ISSUE THREAD

**Issue by mikhail-barg** [Microsoft/CNTK#730]
I've tried running the code based on `CSEvalClient` example. I've installed latest  Microsoft.Research.CNTK.CpuEval-mkl v1.6.0 via NuGet, and referenced the dlls from `\support\x64\Debug`.

The code fails on call to `model.CreateNetwork()` with the following message:

> An unhandled exception of type 'Microsoft.MSR.CNTK.Extensibility.Managed.CNTKRuntimeException' occurred in EvalWrapper.dll
> 
> Additional information: configparameters: required parameter missing: unknown:modelPath

Which is rather weird because I'm pretty sure that the `modelPath` was specified properly.
And anyway if I switch to using libs from `\support\x64\Release`, same code works just well. 

So it seems that something is broken in Debug libs.

**Comment by zhouwangzw** [comment]
Hi,

Can you run the original CSEvalClient exmaple with dlls from Debug build?

Are you using a full path or relative path when specifying modelPath? Currently, the modelPath has to be specified as a full path. In the CSEvalClient example, the relative path is combined with the current working directory to build the full path. 

In addition, mapped drive is not supported for modelPath. Another clarification is that CreateNetwork  must be passed a string containing the "modelPath=<filepath>"  attribute specifying the model's file path.

**Comment by mikhail-barg** [comment]
Hi, Thanks for the answer.

> Can you run the original CSEvalClient exmaple with dlls from Debug build?

Not right now, sorry.

> Are you using a full path or relative path when specifying modelPath? Currently, the modelPath has to be specified as a full path. In the CSEvalClient example, the relative path is combined with the current working directory to build the full path.

I've tried a lot of variants. And yes, I followed the way it's described in example.

> In addition, mapped drive is not supported for modelPath. Another clarification is that CreateNetwork must be passed a string containing the "modelPath=" attribute specifying the model's file path.

No the drive is not mapped. Yes, I've been passing the "modelPath=" attribute.

Once again, just switching native dlls from `\support\x64\Debug\` to `\support\x64\Release` solved the issue. No code changes were needed.

**Comment by zhouwangzw** [comment]
Hi,

Are you building your application in Release or Debug mode? You should only use Debug dlls for a Debug build, and Release dlls for a Release build.

**Comment by ragnar0K** [comment]
Hello there... I am facing the same issue with Linux...
~/cntk/Examples/Evaluation/CPPEvalClient$ ./test 
LockDevice: Locked GPU 0 to test availability.
LockDevice: Unlocked GPU 0 after testing.
LockDevice: Locked GPU 0 for exclusive use.
terminate called after throwing an instance of 'Microsoft::MSR::CNTK::ExceptionWithCallStack<std::runtime_error>'
  what():  configparameters: required parameter missing: unknown:modelPath

**Comment by zhouwangzw** [comment]
@ragnar0K  Which complier version are you using? on which Linux distribution? 
The CNTK Eval in the LINUX binary drop is currently built with GNU C++ 4.8.4. If you have a different compiler version, you may have the issue you met. In this case, please try with the same compiler version, or use the Docker container as described in the Wiki: [CNTK Docker Containers ](https://github.com/Microsoft/CNTK/wiki/CNTK-Docker-Containers)

**Comment by ragnar0K** [comment]
Hello,
I got the binary indeed...
In any case, here is the g++ --version on my Debian testing:
g++ (Debian 6.1.1-11) 6.1.1 20160802

Strangely enough, I tried using g++ 4.9 and it does not compile (4.8 is not available on testing that I can see)

**Comment by zhouwangzw** [comment]
What kind of error did you get when building with g++ 4.9? If possbile, I would suggest to use Docker to build and run your application. I tried to build CPPEvalClient.cpp using the libeval.so from the 1.7 binary package on Ubuntu 16 with GNU C++ 5.4. The build was successful, but during runtime got the same error as you.

**Comment by ragnar0K** [comment]
Unfortunately I wiped all out and started to install from source (so can't test it), but unfortunately I found another issue... The CUDA toolkit requires g++ 4.8 :)
I think I'll leave this here and I'll definitely come back in a year (the few trainings I've seen had quite an impressive speed) - would be great if you guys can remove some dependencies on the way (particularly for making licensing easier to track - which is why I would refrain from using docker). Wish you a great job ahead and thank you much for helping!

**Comment by ignazwrobel007** [comment]
Hi,

I'm also testing the `CSEvalClient` using the binary version of CNTK (1.7-Win64-GPU on Win7, VS2013)

I tried evaluations using trained models of the MNIST example
Using the model *01_OneHidden* worked fine, 
but after changing to *02_Convolution*, the exception
> CPUMatrix<ElemType>::MultiplyAndWeightedAdd : The inner dimensions of a and b must match.

was thrown (after the call `model.Evaluate(...)`).

BTW, I also (successfully) compiled the source version of CNTK 1.7. 
In this version, `CSEvalClient` references the package
> CNTK.CpuEval-mkl version="1.7.**1**"

which cannot be resolved (not released yet?)

I hope you can help me

**Comment by zhouwangzw** [comment]
@ignazwrobel007 What do you mean by "chaning to 02_Convolution"? There is an sample in CSEvalClient, i.e. EvaluateMultipleModels(), which uses the 02_Convolution model. Did you get the exception when running this sample? Could you please attach the output and call stack for investigation? 

The Nuget 1.7.1 is on its way for publishing,  and should be public avaiable by end of today. By "source version of CNTK 1.7" you mean the lastest master? You should be able to compile it after Nuget 1.7.1 is published. 

Thanks.

**Comment by zhouwangzw** [comment]
@ignazwrobel007 I looked into your exception issue. It seems a kind of memory issue between the managed and native boundary. I am still investigating it. 
ps. The Nuget 1.7.1 is published.

Zhou

**Comment by ignazwrobel007** [comment]
@zhouwangzw 
thx for the hint regarding package version 1.7.1, I already could fetch it;
unfortunately it didn't change the result, I'm still getting the error

> CPUMatrix<ElemType>::MultiplyAndWeightedAdd : The inner dimensions of a and b must match.

using `02_Convolution`
I uploaded the exception incl. stacktrace etc. from my own simple test program, which works
with `01_OneHidden`

[test_output.txt](https://github.com/Microsoft/CNTK/files/499413/test_output.txt)

PS: It's great to get so much support here

**Comment by zhouwangzw** [comment]
@ignazwrobel007 Thanks for the log. I am looking into it.

**Comment by zhouwangzw** [comment]
@ignazwrobel007 I think I have found the bug, and we will publish a fix soon. If you are blocked by that, please ping me and I could provide you a hot fix. 
Thanks,

**Comment by ignazwrobel007** [comment]
@zhouwangzw: I'd really like to continue implementing my tests with the convolutional models
I trained, so I'm tempted to come back to your offer; but I wouldn't want to steal your
time

**Comment by zhouwangzw** [comment]
@ignazwrobel007 If you are able to build CNTK from source, you can try with the branch zhouwang/managedeval-memcrash.

**Comment by ignazwrobel007** [comment]
@zhouwangzw I could build the memcrash-branch (after installing CNTK-MKL 2), but I still get the 
error in `CSEvalClient` (nuget package 1.7.1)
Guess I got something wrong - I'll wait for the release

**Comment by zhouwangzw** [comment]
@ignazwrobel007 You have to build CNTK completely from source, e.g. using cntk.sln, and use eval dlls from your binaries, instead of Nuget package. Anyway, we will publish a new version very soon. You might could wait for it.

Thanks and sorry for the inconvenience.

**Comment by ignazwrobel007** [comment]
@zhouwangzw thx for the hint, it worked! (I could run my tests with the convoluational models)

**Comment by zhouwangzw** [comment]
@ignazwrobel007 The new Nuget package 1.7.2 is now available. You can use it, and please let us know if you have any problems.

Thanks.

**Comment by ignazwrobel007** [comment]
@zhouwangzw no problems so far, thx!

**Comment by zhouwangzw** [comment]
Thanks for your patience. I am closing this issue. Please reopen it if needed.
